src/sagemaker/jumpstart/payload_utils.py

```diff
@@ -0,0 +1,134 @@
+        self.bucket = bucket
+        self.region = region
+        self.s3_client = s3_client
+
+    def get_bytes_payload_with_s3_references(
+        self,
+        payload_str: str,
+    ) -> bytes:
+        """Returns bytes object corresponding to referenced s3 object.
+
+        Raises:
+            ValueError: If the raw bytes payload is not formatted correctly.
+        """
+        s3_keys = re.compile(S3_BYTES_REGEX).findall(payload_str)
+        if len(s3_keys) != 1:
+            raise ValueError(f"Invalid bytes payload: {payload_str}")
+
+        s3_key = s3_keys[0]
+        serialized_s3_object = JumpStartS3Accessor.get_object_cached(
+            bucket=self.bucket, key=s3_key, region=self.region, s3_client=self.s3_client
+        )
+
+        return serialized_s3_object
+
+    def embed_s3_references_in_str_payload(
+        self,
+        payload: str,
+    ) -> str:
+        """Embeds s3 references in string payloads."""
+        return self._embed_s3_b64_references_in_str_payload(payload_body=payload)
+
+    def _embed_s3_b64_references_in_str_payload(
+        self,
+        payload_body: str,
+    ) -> str:
+        """Performs base 64 encoding of payloads embedded in a payload.
+
+        This is required so that byte-valued payloads can be transmitted efficiently
+        as a utf-8 encoded string.
+        """
+
+        s3_keys = re.compile(S3_B64_STR_REGEX).findall(payload_body)
+        for s3_key in s3_keys:
+            b64_encoded_string = base64.b64encode(
+                bytearray(
+                    JumpStartS3Accessor.get_object_cached(
+                        bucket=self.bucket, key=s3_key, region=self.region, s3_client=self.s3_client
+                    )
+                )
+            ).decode()
+            payload_body = payload_body.replace(f"$s3_b64<{s3_key}>", b64_encoded_string)
+        return payload_body
+
+    def embed_s3_references_in_json_payload(
+        self, payload_body: Union[list, dict, str, int, float]
+    ) -> Union[list, dict, str, int, float]:
+        """Finds all s3 references in payload and embeds serialized s3 data.
+
+        S3 bucket is assumed to be the default JumpStart content bucket. If no s3 references
+        are found, the payload is returned un-modified.
+        """
+        if isinstance(payload_body, str):
+            return self.embed_s3_references_in_str_payload(payload_body)
+        if isinstance(payload_body, (float, int)):
+            return payload_body
+        if isinstance(payload_body, list):
+            return [self.embed_s3_references_in_json_payload(item) for item in payload_body]
+        if isinstance(payload_body, dict):
+            return {
+                key: self.embed_s3_references_in_json_payload(value)
+                for key, value in payload_body.items()
+            }
+        raise ValueError(f"Payload has unrecognized type: {type(payload_body)}")
+
+    def serialize(self, payload: JumpStartSerializablePayload) -> Any:
```

Comment:
philosophical question: by definition, shouldn't a method called `serialize` return ` str` ?

Reply:
technically it should be `Union[bytes, str]`. The string valued payloads automatically receive utf-8 encoding.